app/controllers/api/visits_controller.rb

```diff
@@ -49,20 +49,36 @@ def create
     end
 
     def show
-      @visit = Visit.find(params[:id])
+      @visit = load_visit
     end
 
     def destroy
-      @visit = Visit.find(params[:id])
-      if @visit.visitor_can_cancel_or_withdraw?
-        @visit.visitor_cancel_or_withdraw!
+      if visitor_cancellation_response.visitor_can_cancel?
+        visitor_cancellation_response.cancel!
+      elsif visitor_withdrawal_response.visitor_can_withdraw?
+        visitor_withdrawal_response.withdraw!
       end
 
+      @visit = load_visit
       render :show
     end
 
   private
 
+    def visitor_cancellation_response
+      @_visitor_cancellation_response ||=
+        VisitorCancellationResponse.new(visit: load_visit)
+    end
+
+    def visitor_withdrawal_response
+      @_visitor_withdrawal_response ||=
+        VisitorWithdrawalResponse.new(visit: load_visit)
+    end
+
+    def load_visit
+      @_visit ||= Visit.find(params[:id])
```

Comment:
having ```@_visit``` and ```@visit``` is redundant. We could get rid of ```@_visit```
Also ```@_xxxx``` is not consistent with the rest of the code base, instance variables are already private and whenever accessed outside a class they have to be accessed via a method.